### ISSUE TITLE
jest :  SecurityError: localStorage is not available for opaque origins

### ISSUE THREAD

**Issue by kitmoovup** [jsdom/jsdom#2304]
When I run a `jest` with my test cases. It shows the following error when I upgrade the package. In my test cases, there is no localStorage is used. How can I fix this problem?

```
 SecurityError: localStorage is not available for opaque origins
      
      at Window.get localStorage [as localStorage] (node_modules/jsdom/lib/jsdom/browser/Window.js:257:15)
          at Array.forEach (<anonymous>)


```

**Comment by domenic** [comment]
It seems likely you are not setting a URL in your tests, but you, or perhaps Jest, are accessing `window.localStorage`. The Jest maintainers may know more about the best fix, but I hear there is a way to set URLs in your Jest tests?

**Comment by gokulkrishh** [comment]
@ben-mckernan Hi, What's the URL you gave to fix this ??

**Comment by ben-mckernan** [comment]
@gokulkrishh I set it to `http://localhost/` but any valid URL seems to work.

**Comment by domenic** [comment]
I guess this is probably Enzyme-specific, because Enzyme does [the thing that the jsdom docs explicitly warn not to do](https://github.com/jsdom/jsdom/wiki/Don't-stuff-jsdom-globals-onto-the-Node-global)? Not too surprising that broke on a minor release, unfortunately :(

**Comment by justinkchen** [comment]
This broke for me too :(

**Comment by gokulkrishh** [comment]
@ben-mckernan Thanks 👍

**Comment by miamollie** [comment]
Also experiencing this.

 https://github.com/jsdom/jsdom/blob/23d67ebec901b3471b84e63f58a96b51a36f3671/lib/jsdom/browser/Window.js#L255

**Comment by domenic** [comment]
("+1" comments will be marked as spam; emailing everyone in the issue thread is not helpful.)

**Comment by miamollie** [comment]
Apologies. Just trying to help.

Can confirm adding `testURL` to jestConfig as @ben-mckernan  suggested did fix it.

And we are using Enzyme too, if that helps confirm your intuition.

**Comment by bencefr** [comment]
For electron app test I just set it to "file:/" which also works.

**Comment by gokulkrishh** [comment]
@miamollie I added `testURL` as per @ben-mckernan suggestion. Due to that my some other test files are failing. Just FYI. See if its works for you. Your test cases might be diff from mine (TestURL might work for you).

**Comment by ben-mckernan** [comment]
@domenic I am only using jest. So I'm not sure that it is an enzyme issue.

**Comment by miamollie** [comment]
@gokulkrishh Yeah, same, it stopped the localStorage security error but made some other tests fail.

**Comment by kitmoovup** [comment]
@ben-mckernan solution fixed it. Thanks!

**Comment by DcsMarcRemolt** [comment]
@ben-mckernan I'm using jest in an angular setup (with jest-preset-angular), same bug, same solution. So it is not an enzyme problem.

**Comment by csvan** [comment]
Seems like Jest needs to change the default value of `testURL` for this to be mitigated (currently it is `about:blank`.

**Comment by gokulkrishh** [comment]
@DcsMarcRemolt I was just debugging the issue. `Jest` is using a dependency module called `jest-environment-jsdom` in its package.json --> "jsdom": "^11.5.1" caret(^) which means npm will installed jsdom as `11.12.0`. So it broke for most of the users. Issue is already created in jest and linked here. Watch out for it.

**Comment by Zirro** [comment]
I'll reopen this issue to give it more visibility. See [the first comment](https://github.com/jsdom/jsdom/issues/2304#issuecomment-408320484) from Domenic for a solution.

**Comment by HenrikHarmer** [comment]
Adding the following to the jest config in package.json: 
` "testEnvironment": "node"`
solves the problem for me.
Credit to @blu3printchris for helping to solve this.

**Comment by gabemeola** [comment]
I can no longer overwrite and mock the JSDOM implementation on localStorage since this update.

**Comment by domenic** [comment]
Thanks, but as of this time I don't believe that's the best path forward. It isn't a breaking change for jsdom to implement new web platform features; that's a minor version.

It's possible for dependents, even widely-used dependents, to write unfortunate code that detects new features and throws an exception if they appear. I hope we can all agree that even if such dependents exist, jsdom should not forever give up on adding new features in minor releases.

The situation here is not quite that drastic, but seems to be similar. I haven't seen anyone track down the offending Jest code yet, so it's hard to say exactly, but it appears they're performing an operation which is explicitly broken by the normal course of jsdom development, and which would not work in a web browser that implements localStorage.

I realize this is a hard situation for those of you impacted, by no fault of your own, by an unfortunate interaction regarding how your direct dependency is (ab)using your indirect dependency. But I'm hopeful this can be addressed at the right level, of fixing Jest's buggy code, instead of removing a useful feature from all jsdom users due to one dependent's bugs.

**Comment by razor-x** [comment]
Thanks for clarifying. I agree that if the issue is from using jsdom in an unsupported way as described above, then it's not a technically a semver violation and jest should release a patch.

**Comment by oliviertassinari** [comment]
My workaround, use a **blacklist** when looping over all *jsdom* properties and adding them to the global.
<details>
  <summary>The code</summary>

```js
const { JSDOM } = require('jsdom');
const Node = require('jsdom/lib/jsdom/living/node-document-position');

// We can use jsdom-global at some point if maintaining these lists is a burden.
const whitelist = ['HTMLElement', 'Performance'];
const blacklist = ['sessionStorage', 'localStorage'];

function createDOM() {
  const dom = new JSDOM('', { pretendToBeVisual: true });
  global.window = dom.window;
  global.Node = Node;
  global.document = dom.window.document;
  // Not yet supported: https://github.com/jsdom/jsdom/issues/317
  global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });
  global.navigator = {
    userAgent: 'node.js',
  };

  Object.keys(dom.window)
    .filter(key => !blacklist.includes(key))
    .concat(whitelist)
    .forEach(key => {
      if (typeof global[key] === 'undefined') {
        global[key] = dom.window[key];
      }
    });
}

module.exports = createDOM;
```
</details>

---

> Don't stuff jsdom globals onto the Node global

Well, I pass for now. I need the tests to run in jsdom and in real browsers. It's the simplest approach I can think of, it has been working for years. I don't see the same potential in the suggested alternatives.

**Comment by Zirro** [comment]
> Well, I pass for now. I need the tests to run in jsdom and in real browsers.

If your tests can run inside "real browsers", they can run the same way in jsdom too - just provide the same HTML. By assigning to the global instead you're introducing additional complexity and differences compared to how you would run the test in a browser.

**Comment by srodrigo** [comment]
FYI I'm running into the same problem with Mocha, not Jest, after upgrading `jsdom` to `11.12.0`.

**Comment by khmy2010** [comment]
Hi, I just want to understand why the changes are introduced at the first place.

My use cases are simply to assert a function to ensure I wrote it correctly, it is as simple as:
```javascript
const fib = require('./index');

test('Fib function is defined', () => {
  expect(typeof fib).toEqual('function');
});

test('calculates correct fib value for 1', () => {
  expect(fib(1)).toEqual(1);
});

```

and yet the test result appears to be I just did some big application on React with Redux library and things like that, while the real thing is I just simply test out a function as simple as

```javascript
//index.js, yes, only one line, no react no redux no enzyme 
function add(a, b) {}
```

so my question is why all the hassle to introduce breaking changes while sometimes we just want a test runner that just "works"

**Comment by Zirro** [comment]
@khmy2010 Your code and questions involve Jest more than they do jsdom. I'd suggest you create an issue in [their repository](https://github.com/facebook/jest) instead.

**Comment by khmy2010** [comment]
Same error message with the op and almost everyone here (except I am not
using enzyme). If not related I should open an issue at jest. Sorry about
that.

On 30 Jul 2018 21:32, "Zirro" <notifications@github.com> wrote:

@khmy2010 <https://github.com/khmy2010> Your code and questions involve
Jest more than they do jsdom. I'd suggest you create an issue in their
repository <https://github.com/facebook/jest> instead.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/jsdom/jsdom/issues/2304#issuecomment-408864079>, or mute
the thread

**Comment by jbarco** [comment]
If you're creating your `jsdom` instance, you can pass a custom url as a second parameter:

```
const url = 'http://localhost';
const jsdom = new JSDOM('<!doctype html><html><body></body></html>, { url });
```

**Comment by khakulov** [comment]
This broke existing infrastructure. This should be major release 12.0.0 and not minor release 11.12.0. Minor releases should not break existing code.

**Comment by Zirro** [comment]
Domenic explains [in this comment](https://github.com/jsdom/jsdom/issues/2304#issuecomment-408520021) why this was not seen as a breaking change. The code from dependent packages which does [things](https://github.com/jsdom/jsdom/wiki/Don't-stuff-jsdom-globals-onto-the-Node-global) we've discouraged for years will hopefully be fixed by their maintainers soon.

**Comment by j-te** [comment]
Additionally to the fix above, I had to add this to the root of test config: `--env=jsdom`

**Comment by odrzutowiec** [comment]
Adding the following to my jest.config.js
```
testURL: 'http://localhost',
```
solved the problem. Thanks!

**Comment by SimenB** [comment]
Do you think it makes sense from Jest's perspective to change from the default `about:blank` to e.g. `localhost`?

**Comment by odrzutowiec** [comment]
Not sure if I should am an authority in this subject, but form the developer point of I would say localhost makes more sense, about:blank is a case that is never a reality. We are testing our apps that  virtually never have about:blank url

**Comment by csvan** [comment]
The Jest team has added a smarter default value for `testURL`: https://github.com/facebook/jest/pull/6792

**Comment by TimothyGu** [comment]
@SimenB I would say that's a good idea.

**Comment by djandrzej** [comment]
This is a breaking change :/. I highly recommend reverting it and releasing 11.13.0 without this change and 12.0.0 with it.

**Comment by mrobrian** [comment]
Getting the same error as the OP, but in my situation it has been causing a LOT of problems. We lock down the versions of packages we use, and we can only change them at the beginning of a release cycle. We are currently toward the end of the release cycle, so our developer environments have the versions of all the packages and their dependencies as of about a month ago, but when we had the build server create a build, it grabbed the current version of all packages. So while all tests pass locally, they all fail on the build server.

We use the "setupTestFrameworkScriptFile" option in the jest config file to do some setup, including polyfills for (among other things) localStorage and sessionStorage (since we use both in our app). It's basically `window.localStorage = window.localStorage || { ... }`, and the same for sessionStorage, where `...` is a bunch of mock functions. Now none of that is working, even if I change it to always override the default (`window.localStorage = { ... }`).

Additionally, we have unit tests that specifically test for things like `sessionStorage.getItem` being called, but after setting "testURL" to "http://localhost", as recommended above to resolve the localStorage errors, those all fail. Even though we have `window.sessionStorage.getItem = jest.fn();`, doing a subsequent `expect(window.sesssionStorage.getItem).toHaveBeenCalled()` fails saying that it's not a mock function.

While I agree that the addition of a feature is a minor version change, and not a breaking change, when it's something that is a standard part of browsers, and the new implementation seemingly can't be overridden or mocked, that _is_ a breaking change.

The only solution I have found to my problem is to add jsdom to my package.json and specify a version of 11.11.0. This is not ideal, and will cause additional work later when we upgrade packages again, but at lease for now it gets us unblocked.

**Comment by domenic** [comment]
If you have code, e.g. mocking code, which works in browsers but not in jsdom, file a new issue following the issue template and we can investigate. To my knowledge localStorage in jsdom is exactly as mockable as it is in browsers.

I do suggest not running different versions on your build server as your developers are running.

**Comment by mrobrian** [comment]
"I do suggest not running different versions on your build server as your developers are running."

While that sounds good in theory, unless every package developer stops using "^" when specifying dependency versions, or we commit the thousands of folders in our node_modules folder, that's never going to happen. From one developer machine to the next there is likely going to be slight differences in some of the dependencies a level or two down. All I can do is specify the exact versions of the packages that are direct dependencies of our application.

**Comment by domenic** [comment]
https://docs.npmjs.com/files/package-locks

**Comment by Bessonov** [comment]
Fully agree with @mrobrian about "^". Craziest thing ever. npm should remove this way to describe dependencies.

@domenic workaround with `packge-lock.json` and `yarn.lock` is broken by design. Maybe it helps with versions, but then create another problems with code-review and merging and often you need to recreate this file by deletion. It's OK if dependencies updated seldom, but we update very often.

Therefore our `.npmrc`:
```
save-exact = true
package-lock = false
```

**Comment by stinkstiefel** [comment]
it helped me adding this new lines to the package.json:

```
"jest": {
    "verbose": true,
    "testURL": "http://localhost/"
  },
```

**Comment by p8ul** [comment]
If you are using jsdom, make sure you include url 

const dom = new JSDOM(``, {
  url: "https://example.org/",
});

**Comment by qwertie** [comment]
Would the jest devs like to comment on why `"testEnvironment": "node"` is now required for CLI projects to avoid the `localStorage` error? Is it a bug?

**Comment by Zirro** [comment]
This isn't the right place to ask questions for Jest developers - jsdom is a separate project. That said, the comments above already contain the answers to your questions.

**Comment by FrankFang** [comment]
package.json

```
   ...
  "jest": {
    "testEnvironment": "node",
    "roots": [
      "test/javascript"
    ]
  },

```

It works for me.

**Comment by mica16** [comment]
@p8ul was right, don't forget to specify "http://localhost/" (url by default since Jest 23.5.0, see #6792) : 

```
const dom = new JSDOM(``, {
url: "http://localhost/",
});
``` 

The whole works for me. 
**Not** even need to add: 
``` 
"testEnvironment": "node"
```

**Comment by csvan** [comment]
jest 23.5.0 now includes a fix for this, so the workarounds should no longer be needed:

https://github.com/facebook/jest/issues/6766#issuecomment-412516712

**Comment by SeanDunford** [comment]
Similar to @mica16 https://github.com/jsdom/jsdom/issues/2304#issuecomment-412663502

```
const dom = new JSDOM(``, {
  url: "http://localhost",
});
```

Was the only change we needed to make to avoid this error. 

We are using mocha/enzyme. No jest included in our test suite.

**Comment by fugroup** [comment]
Setting `--env node` on the command line works too.

**Comment by tirumalavasu** [comment]
> @gokulkrishh I set it to `http://localhost/` but any valid URL seems to work.

"location.href"  would be nice then. :)

**Comment by geoffreyyip** [comment]
@domenic I recommend updating your comment at the top to say that the Jest maintainers fixed this bug in version 23.5.0: https://github.com/facebook/jest/issues/6766#issuecomment-412516712

**Comment by FawzyMokhtar** [comment]
I commit setting the `testURL` for jest-config to `http://localhost` workes.

**Comment by domenic** [comment]
Since this has been fixed in Jest for some time, let's try closing it and seeing if we end up with lots of duplicates, or if things have settled down enough.

**Comment by Shreya1417** [comment]
"jest": {
    "verbose": true,
    "testURL": "http://localhost/"
  } 
Add this code snippet in package.json file. 
It worked for me.

**Comment by christian-draeger** [comment]
you can also add this to the affected test if jsdom is not needed there

``` javascript 1.6
/**
 * @jest-environment node
 */

import { Matchers, Pact } from "@pact-foundation/pact";

it('contract test', () => {
    expect(2 + 2).toBe(4);
});
```

**Comment by mscottx88** [comment]
I tested both options:

1) Add this to the top of a test file:

```javascript
/**
 * @jest-environment node
 */
```

2) Add this stanza to the package.json:

```javascript
"jest": {
    "testURL": "http://localhost/"
  }
```

Both options worked.

**Comment by fbaholli** [comment]
I had it work by adding: 
```
"jest": {
    "verbose": true,
    "testURL": "http://localhost/"
  }
```

**Comment by haiphu** [comment]
> @gokulkrishh I set it to http://localhost/ but any valid URL seems to work.

I am a newbie. Can you tell me exactly where in the Jest.config.js that I set the testURL?

**Comment by gokulkrishh** [comment]
@haiphu Anywhere in `jest.config.js`

```json
{
"testURL": "http://localhost/"

// Your other config
}
```

**Comment by tylerreece22** [comment]
I added the below to my package.json and it works fine now :)

```
  "jest": {
    "testURL": "http://localhost/"
  },
```

**Comment by anshupitlia** [comment]
> @gokulkrishh I set it to `http://localhost/` but any valid URL seems to work.

Even after setting this, it doesn't work for me

**Comment by dosentmatter** [comment]
Not sure why, but my error was caused by having different typescript versions.

I have a mono-repo setup. All packages had `typescript@^3.3.3` in their package.json. I added a new package and installed the latest `typescript@^3.5.3`. When I ran tests in the existing packages, I got this error.

If I move all packages to the same version - either `typescript@^3.3.3` or `typescript@^3.5.3`, the error goes away.

**Comment by mlcohen** [comment]
@tylerreece22 @gokulkrishh Worked for me! 😄

For those wondering what jest's `testURL` configuration directive does, see https://jestjs.io/docs/en/configuration#testurl-string

**Comment by Tech-Nomad** [comment]
For those who sets the options on jsdom directly (when using Mocha for example). Put this in your setup.js:
let jsdom = require('jsdom-global')(
    undefined,
    {
        url: "http://localhost"
    }
);

**Comment by DavideDaniel** [comment]
Strangely.. I have no jsdom being used.. I'm using jest for testing some node only packages yet this error started blocking CI when upgrading jest versions to latest within `^11` range.. I'm sure other folks are seeing similar issues.. none of the so far recommended changes seem to fix it

**Comment by lecstor** [comment]
if you're not using `jsdom` then you'll want to set the jest config property `testEnvironment` to `node`. (no need to touch `testURL`)
https://jestjs.io/docs/en/configuration#testenvironment-string

**Comment by catalin-luntraru** [comment]
For anyone looking for the actual fix, it's `new JSDOM('', { url: 'https://localhost' })`

**Comment by pavelsust** [comment]
Try using this in your package.json file

 "jest": {
    "verbose": true,
    "testURL": "http://localhost/"
  }

**Comment by sam17896** [comment]
Why setting url doesn't work for me... i am using react-native.. is there anything else that i am missing?

**Comment by zuccha** [comment]
> Why setting url doesn't work for me... i am using react-native.. is there anything else that i am missing?

We had the same issue. In our application we had this code
```
const { JSDOM } = require('jsdom');
const jsdom = new JSDOM('<!doctype html><html><body></body></html>');
```

Turns out we had to add the URL to the JSDOM constructor
```
const { JSDOM } = require('jsdom');
const jsdom = new JSDOM('<!doctype html><html><body></body></html>', {
  url: 'http://localhost/',
});
```

That fixed the issue.

**Comment by Flouwrian** [comment]
> > Why setting url doesn't work for me... i am using react-native.. is there anything else that i am missing?
> 
> We had the same issue. In our application we had this code
> 
> ```
> const { JSDOM } = require('jsdom');
> const jsdom = new JSDOM('<!doctype html><html><body></body></html>');
> ```
> 
> Turns out we had to add the URL to the JSDOM constructor
> 
> ```
> const { JSDOM } = require('jsdom');
> const jsdom = new JSDOM('<!doctype html><html><body></body></html>', {
>   url: 'http://localhost/',
> });
> ```
> 
> That fixed the issue.

This worked for me thank you so much! putting the url in the jest config doesnt seem to work with react-native. putting the url in de jsdom constructor did the trick.

**Comment by gustawdaniel** [comment]
Update jest from 22 to 26 fixed problem.

**Comment by Sroy0202** [comment]
simply use the latest version of jest. currently I'm using 26.5.0 in year 2020 and my problem is solved

**Comment by adi518** [comment]
> Try using this in your package.json file
> 
> "jest": {
> "verbose": true,
> "testURL": "http://localhost/"
> }

No need, `testURL` is set to that by default. It doesn't solve the issue with Jest `26.x`. I had to do what @zuccha did to get around it.

**Comment by Michiel87** [comment]
What if you can't set url to localhost? I've set the url to path so it can load other static files. 
See [sandbox](https://codesandbox.io/s/nice-haibt-2lehn?file=/src/index.js).

**Comment by Maikel-Nait** [comment]
Hi all. I'm actually starting to suffer this problem after upgrading to latest Jest **28.0.0-alpha.7**.  Rolling back to current **27.5.1** works fine.

I've noticed that on Jest 28.x , parameters to the **testEnvironment** have to be passed on a **testEnvironmentOptions** object.

Particularly, the old `testURL: 'http://localhost'` have to be passed as `testEnvironmentOptions: { url: 'http://localhost' }`
But no matter using the old `testURL` or the new `testEnvironmentOptions.url` approach, I'm always getting the famous error:

> SecurityError: localStorage is not available for opaque origins

I'm using `testEnvironment: 'jsdom'`, because I'm testing some React (non-native) application.

Any ideas ?  Am I doing something wrong ? Does version 28 have issues configuring the test URL for jsdom environments ?
(switched to version 28, because I want to give a try at the new resolver supporting the package.json **exports**)

**Comment by SimenB** [comment]
You should report that as a bug to Jest, not here. That said, Jest defaults to localhost, and have done so for years. But you shouldn't install jsdom yourself, it comes with `jest-environment-jsdom`.

Regardless, if it worked with v27 and doesn't with v28 that should be reported (with a minimal reproduction) to Jest 🙂

**Comment by Maikel-Nait** [comment]
Thanks, I'm actually inspecting the code, and can't see anything wrong, but will report on jest Github as well. Thanks

**Comment by chrisbobbe** [comment]
> The root cause is often libraries looping over all jsdom properties and adding them to the global; even if you never use localStorage in your tests, some library or test framework you are depending on is "using" it in this manner.

This is happening in `jest-runtime` at https://github.com/facebook/jest/blob/3390ec4ef6a1b93afa816655f5c1f0605066b15a/packages/jest-runtime/src/index.ts#L1165. I'm opening a Jest issue now.

**Comment by bugzpodder** [comment]
i got this error and it was because i had jest 28 and jest-environment-jsdom 26.  upgrading the latter to v28 and it was fixed

**Comment by C-D-Lewis** [comment]
Not using Jest - fix for me using `browser-env` with Mocha was to pass options [as documented](https://github.com/lukechilds/browser-env#usage):

```js
const browserEnv = require('browser-env');

browserEnv({ url: 'http://localhost' });
```

Hopefully this saves people in the same boat as me (seemingly the road less travelled) some time!

**Comment by yuanlllshuai** [comment]
1. yarn add --dev jsdom-global
2. require('jsdom-global')()

**Comment by lisliefor** [comment]
> 



> Hi all. I'm actually starting to suffer this problem after upgrading to latest Jest **28.0.0-alpha.7**. Rolling back to current **27.5.1** works fine.
> 
> I've noticed that on Jest 28.x , parameters to the **testEnvironment** have to be passed on a **testEnvironmentOptions** object.
> 
> Particularly, the old `testURL: 'http://localhost'` have to be passed as `testEnvironmentOptions: { url: 'http://localhost' }` But no matter using the old `testURL` or the new `testEnvironmentOptions.url` approach, I'm always getting the famous error:
> 
> > SecurityError: localStorage is not available for opaque origins
> 
> Interestingly enough, it also seems Jest 28.x is setting jsdom's url to `http://localhost` by **default**, so the configuration could be ommited entirely as well.
> 
> I'm using `testEnvironment: 'jsdom'`, because I'm testing some React (non-native) application.
> 
> Any ideas ? Am I doing something wrong ? Does version 28 have issues configuring the test URL for jsdom environments ? (switched to version 28, because I want to give a try at the new resolver supporting the package.json **exports**)
> 
> **jsdom** library used: tried both **18.1.1** and **19.0.0**

I also meet this question, the version of jest and jest-cli as both v29.5, even I set testEnvironment and testEnvironmentOptions, it still doesn't work.

Finally, I roll back those version to 27.5.1, it works. 
Then I try to upgrade them to v29.5, and install jest-environment-jsdom, it also works.
It seems caused by the default version of jest-environment-jsdom is v24 (you can see it in package-lock.js), it can't work well with the newest jest and jest-cli, should keep them in same version.
